src/main/java/org/jetbrains/research/anticopypaster/utils/Flag.java

```diff
@@ -0,0 +1,86 @@
+package org.jetbrains.research.anticopypaster.utils;
+
+import org.jetbrains.research.extractMethod.metrics.features.FeaturesVector;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+public abstract class Flag{
+
+    protected int sensitivity;
+
+    protected List<FeaturesVector> featuresVectorList;
+
+    protected float metricQ1;
+    protected float metricQ2;
+    protected float metricQ3;
+
+    public abstract boolean isFlagTriggered(FeaturesVector featuresVector);
+
+    public Flag(List<FeaturesVector> featuresVectorList){
+        this.featuresVectorList = featuresVectorList;
+        this.metricQ1=0;
+        this.metricQ2=0;
+        this.metricQ3=0;
+    }
+
+    /**
+    Takes a SORTED list and generates/sets the Q1-3 values based on a box plot
+    of those metric values
+     */
+    protected void boxPlotCalculations(ArrayList<Float> data){
+
+        if(data == null || data.size() == 0){
+            metricQ1=0;
+            metricQ2=0;
+            metricQ3=0;
+            return;
+        }
+
+        float q1;
+        if (data.size() % 2 == 0) {
```

Comment:
So I'm not really sure about the logic behind the "% 2 == 0" and why it's there. Not sure why you can't just do "/4". Might be worth a comment or maybe I'm just dumb.

Reply:
Also since you do the exact same thing for all 3 data points it's probably worth it to have it in a method since this is a lot of repeated code making this look way bigger then it needs to be.

Reply:
I did clean up the logic for the even/odd check in my most recent commit, but that check does need to stay there since the logic for box plotting even/odd length arrays is a little different